Learner mode settings

dadish opened the pull request:
Fixes #4510
Fixes #4511
Fixes #4555

### Changes proposed in this Pull Request

* Added basic support for "deep linking" for Sensei Settings page. With this change, if you append section id into the Sensei Settings page url then it will select that section tab. For example: `/wp-admin/admin.php?page=sensei-settings#course-settings` will pre-select (pre-open) the courses section (tab) of the settings.
* Added `Sensei_Course_Theme::get_sensei_theme_customize_url()` method that generates a url that allows admins/teachers to customize the course theme styles. This basically gets the customize url for the latest lesson that is assigned to any course. If that course has no learner mode enabled, then the url opens in preview mode. This one probably needs some decision from the product. Which page should be opened for customizing Learner mode styles? Is it the latest edited Lesson post? Is it the latest edited Course post?
* Added the global **Learner mode** setting. Which overwrites the **Learner mode** setting for the individual courses. We still need design for learner mode setting for the individual course editor page when the global learner mode is enabled. For now it simply says "Learner mode is enabled globally". Which links to the global setting.
* Sensei settings are now exposed into js runtime on the admin. This allows us to detect if the **Learner Mode** is enabled or not on the Course editor page. Maybe also useful in other parts of Sensei.

### Testing instructions

<!--
If the functionality of this PR is behind a feature flag, please include the
following testing step, using the correct name for the feature flag. (If you
aren't sure, just ignore this step)
-->
* Enable the feature flag: `add_filter( 'sensei_feature_flag_course_theme', '__return_true' );`
* Edit any Course and confirm that you can turn on/off the **Learner Mode**. (see video below)
* Confirm that you can preview the course in Learner Mode when Learner mode is enabled for that course.
* Confirm that you can preview the course in Learner Mode when Learner mode is **not** enabled for that course.
* Confirm that you can customize the Learner Mode styles when Learner mode is enabled for that course.
* Confirm that you can customize the Learner Mode styles when Learner mode is **not** enabled for that course.
* In the Sensei Settings page, in Courses tab, confirm that you can enable/disable the Learner Mode for all courses.
* In the Sensei Settings page, in Courses tab, confirm that the "Customize Styles" allows you to customize Learner Mode styles.

<!-- Add the following only if there are new/updated actions or filters. Please provide a brief description of what they do and any arguments they may take. Be sure to also add the "Hooks" label to this PR. -->
### New/Updated Hooks

* `sensei_course_learning_mode_enabled` Filters if a course has learning mode enabled.
  * Parameters:
    * `{boolean} $enabled_for_course`. True if the learning mode is enabled for the course.
    * `{boolean} $enabled_globally`. True if the learning mode is enabled globally.
    * `{int} $course_id`. The id of the course.
  * Returns: `boolean`.


<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video


https://user-images.githubusercontent.com/2578542/147478201-72f66667-1674-4de3-87fc-59547e848212.mp4

LavaGolem commented:
Taking a look

LavaGolem commented on includes/class-sensei-settings.php:
```diff
@@ -840,6 +849,48 @@ private function get_changed_setting_values( $field, $new_value, $old_value ) {
+	 *
+	 * @param array $args The field arguments.
+	 */
+	public function render_learning_mode_setting( $args ) {
```
Small thing, I would move this function to a separate, with HTML it feels a bit out of place here

dadish replied:
I agree, looks a little odd on it's own, because it's the only html in this file. But this class (`Sensei_Settings`) extends the `Sensei_Settings_API`, which is all about rendering the plugin settings markup. This method simply adds an edge case that is not handled by the `Sensei_Settings_API` class.

LavaGolem reviewed (commented):
Good job :D 
- Went through the testing scenarios from your description and it is all working as expected 
- I left some comments, but they are more of a cosmetic nature
- One thing to fix is lining is failing
- And one good to have a thing, maybe adding some tests, if they don't take too much time. To test for example conditional renderings, check if learner mode is enabled since it has multiple ways it can be enabled

dadish commented on includes/course-theme/class-sensei-course-theme.php:
```diff
@@ -245,4 +250,62 @@ public function enqueue_styles() {
+	 *
+	 * @return bool
+	 */
+	public static function is_sensei_theme_preview_mode( $course_id ) {
```
Updated.

dadish commented on includes/course-theme/class-sensei-course-theme.php:
```diff
@@ -245,4 +250,62 @@ public function enqueue_styles() {
+				],
+			]
+		);
+		if ( empty( $result ) || count( $result ) < 1 ) {
```
Nice catch. Updated. 👍

dadish commented:
> * And one good to have a thing, maybe adding some tests, if they don't take too much time. To test for example conditional renderings, check if learner mode is enabled since it has multiple ways it can be enabled

I added some tests. f685f43ad1fac6915b7a779219e914b591df2a25

dadish commented on includes/course-theme/class-sensei-course-theme.php:
```diff
@@ -245,4 +250,62 @@ public function enqueue_styles() {
+		$query_var = isset( $_GET[ self::PREVIEW_QUERY_VAR ] ) ? intval( $_GET[ self::PREVIEW_QUERY_VAR ] ) : 0;
+
+		// Do not allow sensei preview if requested course id does not match.
+		if ( ! $course_id || $query_var !== $course_id ) {
```
Yep. Removed. 👍

yscik commented:
On the copy:

- Since we are not calling the feature *course theme* anymore, I think we can get rid of the *'This will not change your theme'* line.
- The toggle could also say just 'Learning Mode' everywhere without 'Sensei'
- Should the 'Customize Learning Mode' link be hidden when it's turned off?

dadish commented:
> Since we are not calling the feature _course theme_ anymore, I think we can get rid of the _'This will not change your theme'_ line.

Agree. Removed.
 
> The toggle could also say just 'Learning Mode' everywhere without 'Sensei'

Updated. 

> Should the 'Customize Learning Mode' link be hidden when it's turned off?

We thought it would be good for a user to be able to customize and see how it would look like without actually enabling it to the learners.

renatho commented on includes/class-sensei-settings-api.php:
```diff
@@ -1060,6 +1061,19 @@ public function enqueue_field_styles() {
+		wp_enqueue_script( 'expose-sensei-settings' );
+		wp_add_inline_script(
+			'expose-sensei-settings',
+			sprintf( 'window.senseiSettings = %s;', $settings ),
```
We started introducing `window.sensei` in some parts. Maybe we could also use it here, like this:

`'window.sensei = window.sensei || {}; window.sensei.senseiSettings = %s;'`

dadish replied:
Updated. e6bbb663792477a78cb6c9afdf7ca0c9fc9217d0

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,147 @@
+	const firstLesson = useSelect( ( select ) => {
+		const { getBlocks, getBlockAttributes } = select( 'core/block-editor' );
+		const blocks = getBlocks();
+		const courseOutline = blocks.find(
```
I think if we have the outline as a child of another block, like the Conditional Content block, for example, it wouldn't work, right? Maybe `getFirstBlockByName` from `assets/blocks/course-outline/data.js` can help you with that?

dadish replied:
Nice catch! Updated. 2cb0011a6840bd30a5b76473c71a031b97933a92
Thanks for the `getFirstBlockByName` method!

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,147 @@
 
+	const globalLearningModeEnabled =
+		window.senseiSettings.sensei_learning_mode_all || false;
+	const currentPost = useSelect( ( select ) =>
```
Maybe extracting all this part that gets the `customizerUrl` and `previewUrl` would make this component cleaner. It could be a React Hook inside this same file since it's used only here. Does it make sense?

dadish replied:
Done. dbf9b20ea640685ec546bcf62b0ddfe52c5c4fc1

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,147 @@
-				onChange={ setTheme }
-			/>
+			{ globalLearningModeEnabled ? (
+				<>
```
I think since we just have the `<p>` element here, we don't need the Fragment, right? So we could save some indentation. =)

dadish replied:
Yep. Removed.

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,147 @@
+					</p>
+				</>
+			) : (
+				<>
```
Same about the fragment here.

dadish replied:
Removed.

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,147 @@
+const canPreview = ( block ) =>
+	block.name === courseLessonBlockName &&
+	block.attributes.id &&
+	! block.attributes.draft;
```
Just a question: Have you explored if we could make something to preview draft lessons too?

dadish replied:
I did now. 😄 f79540784de9520007e2463e0a00e7224b83baef

I added a support for it. But it looks a bit empty. Course navigation sidebar is empty, because it excludes the unpublished lessons. But we can improve it further in future iterations. LMWYT. See video below


https://user-images.githubusercontent.com/2578542/148183215-8354f762-81a6-4100-99bf-9dbb2540b11f.mp4

renatho replied:
I think we could start with this version and improve in future iterations, as you said. =)

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,147 @@
+								<p className="sensei-course-theme-toggle__label">
+									{ __( 'Learning mode', 'sensei-lms' ) }
+								</p>
+								<p className="sensei-course-theme-toggle__description">
```
From this part, I think it would be nice to be separate to be more semantic. Otherwise, everything would be part of the `<label>`.

ToogleControl has a [help prop](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/toggle-control) that we could use for the rest of the content. Also, probably we don't need the `<p className="sensei-course-theme-toggle__label">`.

dadish replied:
Thanks for the tip. Updated. 9cd0c14eacade6e704cc491624e77b48865258e2

renatho replied:
Should we also remove the `__label` class from the CSS?

dadish replied:
Yep, forgot about that. Removed the CSS. 0ca13acf831cf54eb97e6f8648ea8a77bbc7b45f

renatho commented on assets/js/settings.js:
```diff
@@ -1,44 +1,35 @@
-
-		sensei_log_event( 'settings_view', { view: toShow } );
-
+	$senseiSettings = $( '#woothemes-sensei.sensei-settings' );
```
Good refactor and good job handling the URL! =)

renatho commented on includes/course-theme/class-sensei-course-theme-option.php:
```diff
@@ -140,7 +144,11 @@ public function should_use_sensei_theme() {
 	 * @return bool
 	 */
-	public function has_sensei_theme_enabled( int $course_id ) {
+	public static function has_sensei_theme_enabled( int $course_id ) {
```
I think we can take the opportunity of this PR, and also add a filter here for users that want to create exceptions here. Could we add it in this PR?

Just notice that since we're using the "learning mode" copy now, it would be nice to use the filter with this copy too.

dadish replied:
Good point! Added a `sensei_has_learning_mode_enabled` hook. 8e6686334743c191ddf589e2dd0f4bc9318e0bd2

renatho replied:
I think you also added a `use phpDocumentor\Reflection\Types\Boolean;` as part of this PR. Was it accidental or on purpose?

Also, I think something like `sensei_course_learning_mode_enabled` would be a little more intuitive for who is reading the filters. Although the method is `has_sensei_theme_enabled` (a check), the purpose while using the filter would be to enable the learning mode. Does it make sense? cc @donnapep since she's thinking about hook names standard. Maybe she has an insight into that.

Also, it would be nice to add the hook to the PR description too. ;)

dadish replied:
> I think you also added a use phpDocumentor\Reflection\Types\Boolean; as part of this PR. Was it accidental or on purpose?

Nope. Probably my editor auto imported it and I didn't notice. Thanks for spotting it!

> Also, I think something like sensei_course_learning_mode_enabled would be a little more intuitive for who is reading the filters. Although the method is has_sensei_theme_enabled (a check), the purpose while using the filter would be to enable the learning mode. Does it make sense? cc @donnapep since she's thinking about hook names standard. Maybe she has an insight into that.

Also, it would be nice to add the hook to the PR description too. ;)
Updated the name to `sensei_course_learning_mode_enabled`. ade117ee8d8024a2844777e1255df8c7942cff5b. Also updated the description.

renatho commented on includes/rest-api/class-sensei-rest-api-lessons-controller.php:
```diff
@@ -139,4 +140,13 @@ protected function add_additional_fields_to_object( $prepared, $request ) {
+	 *
+	 * @param array $args The args.
+	 */
+	public function filter_lesson_meta_args( $args ) {
```
I think this filter isn't doing anything, right? Or am I missing something?

dadish replied:
Yeah. This one is redundant. It's a leftover from my first approach for determining the preview/customize urls for learning mode. Removed! 👍

renatho reviewed (commented):
Good job! Adding a few comments. ;)

dadish commented:
@renatho @yscik Need another round of reviews, please.

renatho commented on assets/blocks/single-course-style-editor.scss:
```diff
@@ -1,2 +1,3 @@
 @import 'course-outline/course-outline-editor';
-@import 'course-progress-block/course-progress-editor'
+@import 'course-progress-block/course-progress-editor';
+@import '../css/sensei-course-theme/course-settings-editor';
```
About this file, since this folder only handles blocks, I think we could import it in another place.
Maybe we could add it as part of `assets/css/course-editor.scss` that I added [here](https://github.com/Automattic/sensei/pull/4599/commits/cf6afb47e6bda49823d98bc97b4813851701767d) (not merged yet). WDYT?

dadish replied:
Removed that CSS file altogether. I had setup styles for settings originally in anticipation with very custom designs, but now it looks like those were not needed so much. 👍

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,152 @@
+			{ customizerUrl && (
+				<p className="sensei-course-theme-toggle__customize">
+					<a
+						href={ customizerUrl }
```
Similar to the preview, maybe we could also open the customizer in target blank by default. WDYT?

dadish replied:
The way I setup them is the way it works in WordPress by default. The regular post preview link in WordPress opens in a separate tab and the customize link opens in the same tab. So I replicated the same behavior.

I think the reason for this behavior in WordPress is that in the customizer mode the user can see the changes they are making directly. But in preview mode you need to switch to editor to make changes and then switch to preview to see the updates. It makes more sense to keep the preview and editor in two separate tabs. But there is no benefit in keeping the customizer in a separate tab, because you have both the changes and the outcome at the same place.

But let me know if you think it's better if we open the customizer in a separate window and I will update it. 👍

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,152 @@
+					label={ __( 'Learning mode', 'sensei-lms' ) }
+					help={
+						<>
+							<p className="sensei-course-theme-toggle__description">
```
I was thinking that we could remove this `p` and the CSS styles for this. It will be a little different from the design, but it will match the WP standards, so it gets the same as other helper texts in the sidebar. Does it make sense?

dadish replied:
Absolutely! I removed that `p` element. 5f01634d45d0f59b5977e7e8560e13b0612d4a2f

But, let me know how do you want to render the preview link. It could either be part of the helper text or it could show up separately to make sure it's visible. Right now it's rendered separately like in the option 2.

| 1 | 2 |
|---|---|
|<img width="270" alt="Screen Shot 2022-01-11 at 12 25 07 PM" src="https://user-images.githubusercontent.com/2578542/148907363-dbdc626b-f20b-4c71-8c47-b840a1f49686.png">|<img width="272" alt="Screen Shot 2022-01-11 at 12 26 38 PM" src="https://user-images.githubusercontent.com/2578542/148907371-e644357c-370b-4c39-89fe-d2053f8fc56b.png">|

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,152 @@
+								) }
+							</p>
+							{ previewUrl && (
+								<p className="sensei-course-theme-toggle__preview">
```
Is there some reason for the preview to be part of the help text and the customizer to be out of this? If not, I think we could add both links after the toggle. It could also be a `ul` to be more semantic. WDYT?

dadish replied:
No reason anymore. It was part of the field helper text before so it would properly align with the toggle field label as in the designs (see below). But now, since we are not sticking to designs strictly, I took it out of the `help` prop of `ToggleControl` component.

I put them in the `p` tags instead of `ul>li` because the latter needs additional CSS to align them properly. So it will require some additional CSS file to be created and played with margins and paddings to keep em in `ul>li` tags. But if you don't like the `p` tag, then let me know and I will update them to be in `ul>li` tags. 👍 

#### Original Design
<img width="351" alt="Screen Shot 2022-01-11 at 12 54 48 PM" src="https://user-images.githubusercontent.com/2578542/148911261-3192b49d-f643-46e5-a7a3-356c22718ebf.png">
 
#### Current Implementation
<img width="271" alt="Screen Shot 2022-01-11 at 12 59 16 PM" src="https://user-images.githubusercontent.com/2578542/148912169-72da2395-720b-4c37-ba04-c7ca638f11d5.png">

renatho commented on assets/css/sensei-course-theme/course-settings-editor.scss:
```diff
@@ -0,0 +1,27 @@
+.sensei-course-theme-toggle {
+	margin-bottom: 20px;
```
I noticed we don't have the "sensei-course-theme-toggle" class applied to any place. Maybe after a refactor and we forgot to remove it?
I think we could clear these styles removing this margin, and the `.components-base-control__field`, or fix it if needed.

We probably could also remove the `&__label` as mentioned here: https://github.com/Automattic/sensei/pull/4565#discussion_r781484490

dadish replied:
Removed the styles. 👍

renatho commented on assets/css/sensei-course-theme/course-settings-editor.scss:
```diff
@@ -0,0 +1,27 @@
+		margin-bottom: 0;
+	}
+
+	&__note {
```
I also didn't see it in use. Is it something we should remove or I missed something?

dadish replied:
Removed. 👍

renatho commented on includes/class-sensei-settings-api.php:
```diff
@@ -1060,6 +1061,19 @@ public function enqueue_field_styles() {
+		wp_register_script( 'expose-sensei-settings', '', [], \false, \false );
+		wp_enqueue_script( 'expose-sensei-settings' );
+		wp_add_inline_script(
+			'expose-sensei-settings',
```
I think it makes more sense to be tied to the `js/admin/course-edit.js` and the `js/settings.js`. So we just add it when needed, and it's added in order to make sure the variable will be ready to use (about that, it's nice to set the last argument - `$position` - of `wp_add_inline_script` to `before`).

dadish replied:
Made it part of the `js/admin/course-edit.js`. e7ea927f748b6133497b36ca323fc9e463f8b60c

renatho commented on includes/course-theme/class-sensei-course-theme-option.php:
```diff
@@ -122,15 +124,19 @@ public function should_use_sensei_theme() {
+			return true;
+		}
+
+		if ( Sensei_Course_Theme::is_sensei_theme_preview_mode_allowed( $course_id ) ) {
```
I saw that it was a suggestion to rename this method, but while reading it in this context, it doesn't make much sense. I think the previous method name would work better. cc @LavaGolem WDYT?

dadish replied:
So which name do you folks prefer? 😄 Let me know when you decide and I'll change it right away.

renatho replied:
Actually, since it's already under the class `Sensei_Course_Theme`, `is_preview_mode` would probably be enough. WDYT?

dadish replied:
Yep. Updated. 👍  cfe65c8b10eb0d4e8c229590e9c399b8313ca0ea

renatho commented on includes/course-theme/class-sensei-course-theme-option.php:
```diff
@@ -122,15 +124,19 @@ public function should_use_sensei_theme() {
 
-		if ( ! $this->has_sensei_theme_enabled( $course_id ) ) {
-			return false;
+		if ( self::has_sensei_theme_enabled( $course_id ) ) {
```
Maybe adding both conditionals in one here would make it more clear?

```
if ( self::has_sensei_theme_enabled( $course_id ) || Sensei_Course_Theme:: is_sensei_theme_preview_mode_allowed( $course_id ) ) {
```

dadish replied:
Sure, updated. 341e6e5fffca51b3b66d04c3fe72633e64149d9a

renatho commented on assets/js/admin/course-theme-sidebar.js:
```diff
@@ -3,49 +3,152 @@
+
+	let customizerUrl = '';
+	if ( previewUrl ) {
+		customizerUrl = `/wp-admin/customize.php?url=${ encodeURIComponent(
```
This URL loaded the customizer on the correct page for me, but it would be nice if it would open directly in the "Sensei LMS > Course Theme" menu. Have you checked if it's possible?

renatho replied:
I found a link with some [instructions](https://wordpress.stackexchange.com/questions/214473/link-to-specific-customizer-section).
I didn't test it, but I hope it works. =)

dadish replied:
Updated! 🎉 Thanks for the tip. I was actually thinking about doing it. Kinda de-prioritized originally and then forgot about it. Now it works great. 👍 972fe20caaffb980c4897219031e6cd74324d007

renatho reviewed (approved):
Looks good and works well! 🎉🎉🎉

burtrw commented on assets/js/admin/course-theme/course-theme-sidebar.js:
```diff
@@ -3,47 +3,145 @@
-			/>
+			{ globalLearningModeEnabled ? (
+				<p>
+					<a href="/wp-admin/admin.php?page=sensei-settings#course-settings">
```
Good points - I will take a close look when I can play :)
